### PR TITLE
`Development`: Move lecture unit features into the lecture micro service 

### PR DESCRIPTION
**Pull request by MerlinMehmed**
#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] *This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [ ] *Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
- [ ] * not related
### Motivation and Context

This is the first PR that moves functionality from Artemis into the Lecture microservice. It moves the lecture attachment unit resource and its test.

### Description
The PR doesn't add new features, we just want to make sure the old ones are still working in the new architecture.

### Steps for Testing

Unfortunately, it is still not possible to test this PR on a test server. It should be tested locally. 
n order to do it, you need to run:
- JHipster Registry by running `docker-compose -f src/main/docker/jhipster-registry.yml up`
- Gateway application by running `./gradlew :gateway:bootRun`
- Lecture application by running `./gradlew :lecture:bootRun --args='--spring.profiles.active=artemis,dev'`
- Artemis as usual or i.e. `./gradlew bootRun --args='--spring.profiles.active=dev,bamboo,bitbucket,artemis,scheduling' -x webapp` to skip the client code build since it's not needed

Prerequisites:
- 1 Instructor
- 1 Student

0. Open [http://localhost:8089/](http://localhost:8089/)
1. Log in to Artemis
2. Navigate to Course Management -> Lectures -> Units
5. Create a new File Unit
6. Open it, make sure you can access it 
7. Edit the created file unit and make sure the changes are saved
8. Check that you can access it as a student, too

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

### CODE REVIEW

**Review by b-fein — APPROVED** [review]
Code changes: Looks good. Nice that you removed the deprecated `ResponseUtil` usages while moving the service. :+1: 

Manual Test: The same fix as described in https://github.com/ls1intum/Artemis/pull/4424#pullrequestreview-820641904 has been applied. Afterwards adding/editing/deleting of the file unit as instructor and looking at/downloading it as a student worked without issues. :+1:

File: src/test/java/de/tum/in/www1/artemis/AttachmentUnitIntegrationTest.java [hunk 1]
```diff
@@ -16,6 +16,7 @@
 import de.tum.in.www1.artemis.repository.*;
 import de.tum.in.www1.artemis.util.ModelFactory;
 
+@Deprecated // Moved to Lecture microservice. To be removed
```
Comment (julian-christl) [inline src/test/java/de/tum/in/www1/artemis/AttachmentUnitIntegrationTest.java:19]: Remove, see lecture unit PR

File: src/test/java/de/tum/in/www1/artemis/AttachmentUnitIntegrationTest.java [hunk 2]
```diff
@@ -16,6 +16,7 @@
...
 public class AttachmentUnitIntegrationTest extends AbstractSpringIntegrationBambooBitbucketJiraTest {
```
Comment (julian-christl) [inline src/test/java/de/tum/in/www1/artemis/AttachmentUnitIntegrationTest.java:20]: Why did you remove the section in `createAttachmentUnit_asInstructor_shouldCreateAttachmentUnit()` without replacement?

File: lecture/src/main/java/de/tum/in/www1/artemis/lecture/web/rest/AttachmentUnitResource.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        }
+        Lecture lecture = lectureRepository.findByIdWithLectureUnitsElseThrow(lectureId);
+        if (lecture.getCourse() == null) {
+            throw new ConflictException();
+        }
+        authorizationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.EDITOR, lecture.getCourse(), null);
+
+        // persist lecture unit before lecture to prevent "null index column for collection" error
+        attachmentUnit.setLecture(null);
+        attachmentUnit = attachmentUnitRepository.saveAndFlush(attachmentUnit);
```
Comment (julian-christl) [inline lecture/src/main/java/de/tum/in/www1/artemis/lecture/web/rest/AttachmentUnitResource.java:115]: fix codacy issue

**Review by julian-christl — CHANGES_REQUESTED** [review]

File: src/test/java/de/tum/in/www1/artemis/AttachmentUnitIntegrationTest.java (hunk 1 above)
Comment (MerlinMehmed) [reply src/test/java/de/tum/in/www1/artemis/AttachmentUnitIntegrationTest.java:19]: I'll remove the @ Deprecated and will only leave the comment so that it's easy to find later when it's time to remove this file

File: src/test/java/de/tum/in/www1/artemis/AttachmentUnitIntegrationTest.java (hunk 2 above)
Comment (MerlinMehmed) [reply src/test/java/de/tum/in/www1/artemis/AttachmentUnitIntegrationTest.java:20]: The removed section was related to calling /api/attachments endpoint which is not part of the lecture microservice. At first I thought that this can be removed as we want to test the attachmentunit. Now, that I had a second look the creation of the attachment is important in order to test the link between the attachment and attachment unit. So I added it back. But since we cannot call the endpoint I call the attachment repository

File: lecture/src/test/java/de/tum/in/www1/artemis/lecture/AttachmentUnitIntegrationTest.java [hunk 4]
```diff
@@ -82,8 +82,12 @@ public void createAttachmentUnit_asInstructor_shouldCreateAttachmentUnit() throw
         var persistedAttachmentUnit = request.postWithResponseBody("/api/lectures/" + this.lecture1.getId() + "/attachment-units", attachmentUnit, AttachmentUnit.class,
                 HttpStatus.CREATED);
         assertThat(persistedAttachmentUnit.getId()).isNotNull();
+        this.attachment.setAttachmentUnit(persistedAttachmentUnit);
+        var persistedAttachment = attachmentRepository.save(attachment);
```
Comment (julian-christl) [inline lecture/src/test/java/de/tum/in/www1/artemis/lecture/AttachmentUnitIntegrationTest.java:86]: If you don't call the endpoint, you can remove it though. You can see that the assertions are very trivial.
e.g. line 85 and 88 or 84 and 87

File: lecture/src/test/resources/config/application.yml [hunk 5]
```diff
@@ -77,6 +77,7 @@ spring:
         host: localhost
     main:
         allow-bean-definition-overriding: true
+        allow-circular-references: true # TODO: we should deactivate this in the future, however currently securityConfiguration has a circular dependency to itself
```
Comment (julian-christl) [inline lecture/src/test/resources/config/application.yml:80]: Do you have this on your TODO list to refactor it? Otherwise, or possibly either way, please create an issue for that, so it doesn't get forgotten.

Comment (julian-christl) [reply lecture/src/test/resources/config/application.yml:80]: Same for the user-management

File: lecture/src/test/java/de/tum/in/www1/artemis/lecture/AttachmentUnitIntegrationTest.java (hunk 4 above)
Comment (MerlinMehmed) [reply lecture/src/test/java/de/tum/in/www1/artemis/lecture/AttachmentUnitIntegrationTest.java:86]: Yeah, it's not possible to call it. Alright, I will remove it, both are fine for me

File: lecture/src/test/resources/config/application.yml (hunk 5 above)
Comment (MerlinMehmed) [reply lecture/src/test/resources/config/application.yml:80]: I don't think it's a good idea to open an issue for that since the microservices are not on develop yet. The need for this change came from some previous changes applied from the develop branch which included this config in the artemis app as well. If it's solved until I am in the project, I will remove it from the config. If not, I can create an issue once the microservices are on develop.

Comment (julian-christl) [reply lecture/src/test/resources/config/application.yml:80]: The reason I would suggest creating the issue now is because people tend do forget things 😅

Comment (MerlinMehmed) [reply lecture/src/test/resources/config/application.yml:80]: I understand, but I still don't think it makes sense creating an issue for implementation that is not in develop yet.

**Review by julian-christl — APPROVED** [review]
Issues addressed, code looks good.
